jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUForestIsomorphismInspector.java

```diff
@@ -0,0 +1,212 @@
+     *
+     * @param forest1 the first rooted forest
+     * @param roots1 the roots of the first forest
+     * @param forest2 the second rooted forest
+     * @param roots2 the roots of the second forest
+     * @throws NullPointerException if {@code forest1} is {@code null}
+     * @throws NullPointerException if {@code roots1} is {@code null}
+     * @throws NullPointerException if {@code forest2} is {@code null}
+     * @throws NullPointerException if {@code roots2} is {@code null}
+     * @throws IllegalArgumentException if {@code roots1} is empty
+     * @throws IllegalArgumentException if {@code roots2} is empty
+     * @throws IllegalArgumentException if either {@code roots1} or {@code roots2} contain an invalid vertex
+     */
+    public AHUForestIsomorphismInspector(Graph<V, E> forest1, Set<V> roots1, Graph<V, E> forest2, Set<V> roots2){
+        this.forest1 = Objects.requireNonNull(forest1, "input forest cannot be null");
+        this.forest2 = Objects.requireNonNull(forest2, "input forest cannot be null");
+
+        this.roots1 = Objects.requireNonNull(roots1, "set of roots cannot be null");
+        this.roots2 = Objects.requireNonNull(roots2, "set of roots cannot be null");
+
+        if (roots1.isEmpty()){
+            throw new IllegalArgumentException("roots1 cannot be empty");
+        }
+
+        if (roots2.isEmpty()){
+            throw new IllegalArgumentException("roots2 cannot be empty");
+        }
+
+        if (!forest1.vertexSet().containsAll(roots1)){
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+
+        if (!forest2.vertexSet().containsAll(roots2)){
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+    }
+
+    private V getFreshVertex(Graph<V, E> graph){
+        Supplier<V> supplier = graph.getVertexSupplier();
+
+        if (Objects.isNull(supplier))
+            throw new IllegalArgumentException("vertex supplier cannot be null");
+
+        V v = supplier.get();
+        assert !graph.vertexSet().contains(v);
+        return v;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Iterator<GraphMapping<V, E>> getMappings() {
+        GraphMapping<V, E> iterMapping = getMapping();
+
+        if (iterMapping == null)
+            return Collections.emptyIterator();
+        else
+            return Collections.singletonList(iterMapping).iterator();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isomorphismExists(){
+        if (!computed){
+            isomorphicMapping = getMapping();
+        }
+
+        return isomorphicMapping != null;
+    }
+
+    private Pair<V, Graph<V, E>> addDummyRoot(Graph<V, E> forest, Set<V> roots){
+        GraphTypeBuilder<V, E> graphTypeBuilder = GraphTypeBuilder.forGraph(forest);
+        graphTypeBuilder = graphTypeBuilder.weighted(false);
+
+        Graph<V, E> freshForest = graphTypeBuilder.buildGraph();
```

Comment:
Builder methods return the builder in order to do cascade calls
`Graph<V, E> freshForest = GraphTypeBuilder.forGraph(forest).weighted(false).buildGraph();`